File: backend-python/util/keycloak_util.py
import os

import requests


def get_admin_token() -> str:
    url = f'{os.environ.get('OAUTH_ISSUER')}/realms/{os.environ.get('REALM_NAME')}/protocol/openid-connect/token'
    data = {
        'grant_type': 'password',
        'client_id': os.environ.get('CLIENT_ID'),
        'client_secret': os.environ.get('CLIENT_SECRET'),
        'username': os.environ.get('KEYCLOAK_ADMIN_USER'),
        'password': os.environ.get('KECYLOAK_ADMIN_PASSWORD')
    }
    response = requests.post(url, data=data)
    response.raise_for_status()
    return response.json()['access_token']

def get_user_by_id(user_id: str):
    admin_token = get_admin_token()

    url = f'{os.environ.get('OAUTH_ISSUER')}/admin/realms/{os.environ.get('REALM_NAME')}/users/{user_id}'
    headers = {'Authorization': f'Bearer {admin_token}'}

    response = requests.get(url, headers=headers)

    response.raise_for_status()

    return response.json()

File: backend/src/test/kotlin/com/beatgear/backend/util/KeycloakUtil.kt
package com.beatgear.backend.util

import org.apache.http.client.utils.URIBuilder
import org.springframework.boot.json.JacksonJsonParser
import org.springframework.http.MediaType
import org.springframework.util.LinkedMultiValueMap
import org.springframework.web.reactive.function.BodyInserters
import org.springframework.web.reactive.function.client.WebClient
import java.util.*

object KeycloakUtil {

    fun getAccessToken(issuerUrl: String): String {
        val authorizationURI = URIBuilder("$issuerUrl/protocol/openid-connect/token").build()

        val webClient = WebClient.builder().build()
        val formData = LinkedMultiValueMap<String, String>()
        formData["grant_type"] = Collections.singletonList("password")
        formData["client_id"] = Collections.singletonList("test-client")
        formData["client_secret"] = Collections.singletonList("nXuaXbAPEDGw5GcZNowdr2lXRyzaTNoh")
        formData["username"] = Collections.singletonList("test-user")
        formData["password"] = Collections.singletonList("test")

        val result = webClient.post()
            .uri(authorizationURI)
            .contentType(MediaType.APPLICATION_FORM_URLENCODED)
            .body(BodyInserters.fromFormData(formData))
            .retrieve()
            .bodyToMono(String::class.java)
            .block()

        val jsonParser = JacksonJsonParser()
        return "Bearer " + jsonParser.parseMap(result)["access_token"].toString()
    }

}